### PR TITLE
libcontainerd: attach streams before create

### LINKED ISSUE #26371: Failed connection to external log-driver run containers hidden from `docker ps`
**Output of `docker version`:**

```
Client:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.7
 Git commit:   23cf638
 Built:        Fri Aug 19 02:03:02 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.7
 Git commit:   23cf638
 Built:        Fri Aug 19 02:03:02 2016
 OS/Arch:      linux/amd64

```

**Output of `docker info`:**

```
Containers: 1
 Running: 1
 Paused: 0
 Stopped: 0
Images: 158
Server Version: 1.12.1
Storage Driver: devicemapper
 Pool Name: docker-8:2-534994-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 3.1 GB
 Data Space Total: 107.4 GB
 Data Space Available: 5.065 GB
 Metadata Space Used: 5.296 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.142 GB
 Thin Pool Minimum Free Space: 10.74 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 WARNING: Usage of loopback devices is strongly discouraged for production use. Use `--storage-opt dm.thinpooldev` to specify a custom block storage device.
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.133 (2016-08-15)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: host overlay bridge null
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 4.7.2-1-ARCH
Operating System: Arch Linux
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 7.691 GiB
Name: pho
ID: EUUF:K3RP:ZYHG:A233:NLDD:VHX3:JWOE:RCJM:ADBL:IRJ5:WKCU:BIXP
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Username: pho
Registry: https://index.docker.io/v1/
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Physical machine

**Steps to reproduce the issue:**
1. Do not be running a fluentd daemon
2.  Run docker run --rm -ti --log-driver=fluentd busybox

**Describe the results you received:**
docker run fails with:
>docker: Error response from daemon: Failed to initialize logging driver: dial tcp 127.0.0.1:24224: getsockopt: connection refused.
>Error response from daemon: Driver devicemapper failed to remove root filesystem f76e4d2374a62766204669d63d9d4a7650e68e77dc803f47d2cc07779674d60f: Device is Busy

On 1.12.1, `docker ps -a` shows no container is running, BUT the container is running silently in the background 

On another Arch machine running 1.11.1, `docker ps -a` shows a dead container, BUT the container is running silently in the background

**Describe the results you expected:**
The container shows the error response and stops. 

**Additional information you deem important (e.g. issue happens only occasionally):**
This happen with any log-driver that opens a socket and fails (tested with fluentd and splunk, with none of them running)

### PR DESCRIPTION
**Pull request by LK4D4**
Fix #26371

### CODE REVIEW

**Comment by icecrime** [comment]
LGTM, PTAL @tonistiigi!

File: integration-cli/docker_cli_run_test.go [hunk 1]
```diff
@@ -4536,3 +4536,17 @@ func (s *DockerDaemonSuite) TestRunWithUlimitAndDaemonDefault(c *check.C) {
...
 	c.Assert(out, checker.Contains, "[nofile=42:42]")
 }
+
+func (s *DockerSuite) TestRunStoppedLoggingDriverNoLeak(c *check.C) {
+	nroutines, err := getGoroutineNumber()
+	c.Assert(err, checker.IsNil)
+
+	out, _, err := dockerCmdWithError("run", "--name=fail", "--log-driver=fluentd", "busybox", "true")
+	c.Assert(err, checker.NotNil)
+	if !strings.Contains(out, "Failed to initialize logging driver") {
```
Comment (AkihiroSuda) [inline integration-cli/docker_cli_run_test.go]: nit: you can use `c.Assert(out, checker.Contains, ..)`

Comment (LK4D4) [inline integration-cli/docker_cli_run_test.go]: Sure, I tried to look for example in this file and failed to find it. Thanks for pointing out.

Comment (LK4D4) [inline integration-cli/docker_cli_run_test.go]: fixed

**Comment by LK4D4** [comment]
I wonder why fluentd is disabled for windows. ping @jhowardmsft

**Comment by jhowardmsft** [comment]
@LK4D4 - Open an issue an ping me on it. I don't know if it will run on Windows.

**Comment by jhowardmsft** [comment]
@darrenstahlmsft Worth checking if we need a similar change on Windows as I know this is an area you've been looking at.

**Comment by LK4D4** [comment]
@tonistiigi PTAL

**Comment by tonistiigi** [comment]
LGTM

**Comment by LK4D4** [comment]
heh, another timeout :/

File: libcontainerd/container_linux.go [hunk 2]
```diff
@@ -91,7 +91,8 @@ func (ctr *container) start(checkpoint string, checkpointDir string) error {
 	if err != nil {
 		return nil
 	}
-	iopipe, err := ctr.openFifos(spec.Process.Terminal)
+	createChan := make(chan struct{})
+	iopipe, err := ctr.openFifos(spec.Process.Terminal, createChan)
```
Comment (tonistiigi) [inline libcontainerd/container_linux.go]: How about re-wrapping it here instead(with a comment about why we need this hack): 
```
stdin := iopipe.Stdin
iopipe.Stdin = ioutils.NewWriteCloserWrapper(stdin, func() error {
  go func() {
    <-createChan
    stdin.Close()
   }()
  return nil
})
```

File: libcontainerd/client_linux.go [hunk 3]
```diff
@@ -86,7 +86,7 @@ func (clnt *client) AddProcess(ctx context.Context, containerID, processFriendly
 		Rlimits:         convertRlimits(sp.Rlimits),
 	}
 
-	iopipe, err := p.openFifos(sp.Terminal)
+	iopipe, err := p.openFifos(sp.Terminal, nil)
```
Comment (tonistiigi) [inline libcontainerd/client_linux.go]: Actually, this case isn't that different. Only difference is that you can't reproduce it with a logdriver(that issue we could fix with easy validation for containers). If the attach fails in here you will still have a process running without docker having control of the io streams.

File: libcontainerd/container_linux.go (hunk 2 above)
Comment (LK4D4) [inline libcontainerd/container_linux.go]: Sounds good.

**Comment by tonistiigi** [comment]
LGTM

**Review by vieux — APPROVED** [review]
LGTM

**Pull request merged**